packages/reporters/cli/src/CLIReporter.js

```diff
@@ -1,16 +1,149 @@
 // @flow strict-local
 
-import nullthrows from 'nullthrows';
+import type {
+  BundleGraph,
+  LogEvent,
+  ParcelOptions,
+  ProgressLogEvent,
+  ReporterEvent
+} from '@parcel/types';
+import type {RerenderFunc} from 'ink';
+
+import {prettifyTime} from '@parcel/utils';
+import {render} from 'ink';
 import {Reporter} from '@parcel/plugin';
 import React from 'react';
-import {render} from 'ink';
+
+import {getProgressMessage} from './utils';
+import logLevels from './logLevels';
 import UI from './UI';
 
-let ui: ?UI;
-render(<UI ref={u => (ui = u)} />);
+type State = {|
+  progress: ?ProgressLogEvent,
+  logs: Array<LogEvent>,
+  bundleGraph: ?BundleGraph
+|};
+
+let state: State = {
+  progress: null,
+  logs: [],
+  bundleGraph: null
+};
+let rerender: RerenderFunc;
 
 export default new Reporter({
   report(event, options) {
-    nullthrows(ui).report(event, options);
+    let newState = reducer(state, event, options);
+    let uiElement = <UI {...newState} options={options} />;
+    if (rerender) {
+      rerender(uiElement);
```

Comment:
maybe we should just use redux to avoid all this? Instead of re-rendering at the top level each time, we'd render once and dispatch actions with these updates.

Reply:
seems like so much for one component...